### ISSUE TITLE
Improve ansible root mysql password handling.

### ISSUE THREAD

**Issue by adrinux** [opendevshop/devshop#165]
** ORIGINAL TITLE:**
> Ansible fail during mysql role stops upgrade completion.

Following an apparently successful 'sudo devshop upgrade' of a 1.x install (after applying patch from #159) and server restart devshop became unresponsive. This turned out to be because nginx had failed to start due to duplicate directives in "/etc/nginx/conf.d/aegir.conf".  The problem directives are removed by [ansible-role-aegir-nginx](https://github.com/opendevshop/ansible-role-aegir-nginx/blob/master/tasks/main.yml).

So either that ansible role isn't being run during upgrades, or ansible is failing before it reaches that task. About to rerun the upgrade and check.

**Comment by adrinux** [comment]
Ansible fails during the upgrade when playbook-nginx is running, specifically whilst working through the geerlingguy.mysql role: 
```
TASK [geerlingguy.mysql : Ensure MySQL is started and enabled on boot.] ********
ok: [ods-dev.perlucida.uk]

TASK [geerlingguy.mysql : Get MySQL version.] **********************************
ok: [ods-dev.perlucida.uk]

TASK [geerlingguy.mysql : Disallow root login remotely] ************************
failed: [ods-dev.perlucida.uk] (item=DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')) => {"changed": false, "cmd": ["mysql", "-NBe", "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')"], "delta": "0:00:01.036542", "end": "2016-11-14 17:15:38.329775", "failed": true, "item": "DELETE FROM mysql.user WHERE User='root' AND Host NOT IN ('localhost', '127.0.0.1', '::1')", "rc": 1, "start": "2016-11-14 17:15:37.293233", "stderr": "ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)", "stdout": "", "stdout_lines": [], "warnings": []}

RUNNING HANDLER [geerlingguy.nginx : reload nginx] *****************************
	to retry, use: --limit @/usr/share/devshop/playbook-nginx.retry

PLAY RECAP *********************************************************************
ods-dev.perlucida.uk       : ok=31   changed=2    unreachable=0    failed=1
```

**Comment by adrinux** [comment]
I suppose the root of the issue is that the playbook needs to use a password here but doesn't.
```
"ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: NO)"
```
Can't figure out why yet.

**Comment by adrinux** [comment]
At first I thought this might be caused by not running the mysql ansible role as root but it seems that is not the issue. Looks like there is an underlying problem with mysql access as root.
/root/.my.cnf is present
I can run mysql cli with that password by invoking it:
```
mysql -u root -p
```
But simply running ```mysql``` as root doesn't work and throws the very common error seen in the second comment. I wonder if these ansible roles are expecting /root/.my.cnf to define a passwordless socket connection, rather than the password?

**Comment by adrinux** [comment]
The problem appears to be that by default in Ubuntu 14.04 the $HOME environment variable is not set to root's when using sudo. To get the upgrade to work fully we need to use the -H flag[1] when calling sudo:
```
sudo -H devshop upgrade
```
I was unable to install devshop until I added -H to the initial install command – I really should have realised we needed this for upgrade sooner! I had assumed this was a quirk of using LXC containers but I believe it's a change in behaviour brought in with Ubuntu 14.04 (via upstream Debian Squeeze).

For most invocations of sudo the $HOME is irrelevant but because the mysql password is stored in /root/.my.cnf any script requiring access to mysql needs $HOME to be /root, not the $HOME of the user you logged in as (at least the way ansible geerlingguy.mysql is currently written).

You can see whether this affects your host OS by checking $HOME, here I log in with the ubuntu user:
```
echo $HOME
gives: /home/ubuntu

sudo -s
echo $HOME
gives: /home/ubuntu
exit

sudo -H -s
echo $HOME
gives /root
```

To sum up:
- A workaround for anyone experiencing this is to use the -H flag when calling the upgrade command via sudo: ```sudo -H -s devshop upgrade```.
- This _could_ be 'fixed' via a documentation change but I'm not sure if it affects CentOS.
- It could also presumably be fixed by an upstream change in geerlingguy.mysql role to explicitly use /root as $HOME.

[1] See https://www.sudo.ws/man/1.8.18/sudo.man.html for a description of -H.

**Comment by jonpugh** [comment]
Wow, great breakdown, thank you!

This is exactly the problem with all of the MySQL connection issues. This is why I changed all requests to run install.sh as **root* just to make sure we're in the right environment.

We can alter the code for the `devshop upgrade` command to check for the home directory and mysql cfg before doing anything else.

**Comment by jonpugh** [comment]
:point_down:
